src/uct/tcp/tcp_ep.c

```diff
@@ -526,25 +496,39 @@ static ucs_status_t uct_tcp_ep_connect(uct_tcp_ep_t *ep)
 {
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
+    uct_tcp_ep_t *peer_ep  = NULL;
     ucs_status_t status;
-    uct_tcp_ep_t *peer_ep;
 
-    ucs_assert(uct_tcp_ep_is_unconnected(ep));
+    /* It checks that the connect was issued:
+     * - from uct_ep_connect_to_ep(), if the EP created using CONNECT_TO_EP
+     *   method and a local side must connect to a peer
+     * - from uct_ep_create(), if the EP created using CONNECT_TO_IFACE method
+     *   and no an internal EP was created due to connection from a peer */
+    ucs_assert((ep->conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) &&
+               (ep->conn_retries == 0));
 
     uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTING);
-    if (uct_tcp_ep_is_self(ep) ||
-        (ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP) ||
-        ((peer_ep = uct_tcp_cm_get_ep(iface, &ep->peer_addr, ep->cm_id.conn_sn,
-                                      UCT_TCP_EP_FLAG_CTX_TYPE_RX)) == NULL)) {
+
+    if (!uct_tcp_ep_is_self(ep) &&
+        !(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP)) {
```

Comment:
"goto out" here if it's not self or connect-to-ep, since peer_ep is going to be NULL anyway
also can avoid the 'if' in line 552-553 by using extra goto label

Reply:
fixed